src/frontend/src/optimizer/rule/pull_up_correlated_predicate_agg_rule.rs

```diff
@@ -0,0 +1,245 @@
+        let agg: &LogicalAgg = input.as_logical_agg()?;
+        let (agg_calls, group_key, grouping_sets, input, _enable_two_phase) =
+            agg.clone().decompose();
+        // It could be too restrictive to require the group key to be empty. We can relax this in the future if necessary.
+        if !group_key.is_empty() {
+            return None;
+        }
+        assert!(grouping_sets.is_empty());
+        let bottom_project = if let Some(project) = input.as_logical_project() {
+            project.clone()
+        } else {
+            LogicalProject::with_mapping(
+                input.clone(),
+                ColIndexMapping::identity(input.schema().len()),
+            )
+        };
+        let (mut bottom_proj_exprs, _) = bottom_project.clone().decompose();
+        let bottom_project_input = bottom_project.input();
+        let filter: &LogicalFilter = bottom_project_input.as_logical_filter()?;
+
+        // Split predicates in LogicalFilter into correlated expressions and uncorrelated
+        // expressions.
+        let (cor_exprs, uncor_exprs): (Vec<ExprImpl>, Vec<ExprImpl>) = filter
+            .predicate()
+            .clone()
+            .into_iter()
+            .partition_map(|expr| {
+                if expr.has_correlated_input_ref_by_correlated_id(correlated_id) {
+                    Either::Left(expr)
+                } else {
+                    Either::Right(expr)
+                }
+            });
+
+        // Ensure all correlated expressions look like [correlated_input_ref = input_ref]
+        let mut cor_eq_exprs = vec![];
+        for cor_expr in &cor_exprs {
+            if let Some((input_ref, cor_input_ref)) = cor_expr.as_eq_correlated_input_ref() {
+                if cor_input_ref.correlated_id() == correlated_id {
+                    cor_eq_exprs.push((input_ref, cor_input_ref));
+                } else {
+                    return None;
+                }
+            } else {
+                return None;
+            }
+        }
+        let cor_eq_exprs_len = cor_eq_exprs.len();
+
+        let filter = LogicalFilter::create(
+            filter.input(),
+            Condition {
+                conjunctions: uncor_exprs,
+            },
+        );
+
+        // Append `InputRef`s in the predicate expression to be pulled to the project, so that they
+        // are accessible by the expression after it is pulled.
+        bottom_proj_exprs.extend(
+            cor_eq_exprs
+                .iter()
+                .map(|(input_ref, _)| ExprImpl::InputRef(input_ref.clone().into())),
+        );
+
+        let new_bottom_proj: PlanRef = LogicalProject::new(filter, bottom_proj_exprs).into();
+
+        // If there is a count aggregate, bail out and leave for general subquery unnesting to deal.
+        if agg_calls
+            .iter()
+            .any(|agg_call| agg_call.agg_kind == AggKind::Count)
+        {
+            return None;
+        };
```

Comment:
why? is count very special here?

Reply:
Yes, here is the corner case. I didn't come up with an idea of how to deal with it now. If you have some ideas, feel free to share. It is also related to the TPCH Q17.
```sql
create table t (a int, b int);

create table t2 (c int, d int);

insert into t values (1, 2);

flush;

select * from t where t.a > (select count(*) from t2 where b = d);
```

Reply:
When group by is empty, count would return 0 instead of null.

Reply:
Yes, I can not find a way to rewrite it... But it should ok for Q17 because the aggregator is an AVG,  we should choose one to optimize q17
- maintain the AVG and other similar aggregator in plan node and delay their rewriting(currently RW rewrites it when creating the Agg plan node)
- consider the project-agg together in the rule later